File: 0720-longest-word-in-dictionary/0720-longest-word-in-dictionary.java
class Solution {
    class TrieNode{
        TrieNode[] children;
        boolean isEnd;
        char c;
        public TrieNode(){
            children = new TrieNode[26];
        }
    }

    TrieNode root;
    private void insert(String word){
        TrieNode curr = root;
        for(int i = 0; i < word.length(); i++){
            char c = word.charAt(i);
            if(curr.children[c - 'a'] == null){
                curr.children[c - 'a'] = new TrieNode();
            }
            curr = curr.children[c - 'a'];
            curr.c = c;
        }
        curr.isEnd = true;
    }
    StringBuilder result;

    public String longestWord(String[] words) {
        root = new TrieNode();
        result =  new StringBuilder();
        for(String word: words){
            insert(word);
        }
        dfs(root, new StringBuilder());
        return result.toString();
    }

    private void dfs(TrieNode curr, StringBuilder path){
        //base
        if(path.length() >= result.length()){
            result = new StringBuilder(path);
        }

        //logic
        for(int i = 25; i >= 0; i--){
            if(curr.children[i] != null && curr.children[i].isEnd){
                //action
                path.append(curr.children[i].c);
                // recurse
                dfs(curr.children[i], path);
                //backtrack
                path.setLength(path.length()- 1);
            }
        }
    }
}

File: 0720-longest-word-in-dictionary/0720-longest-word-in-dictionary.py
class TrieNode:
    children = []
    isEnd = False
    def __init__(self):
        self.children = [None] * 26
        self.isEnd = False

class Solution:
    result = ""
    root = None    
    def __init__(self):
        self.root = TrieNode()

    def longestWord(self, words: List[str]) -> str:
        self.result = ""

        if words is None or len(words) == 0:
            return self.result

        for word in words:
            self.insert(word, self.root)
        
        self.dfs(self.root, [])

        return self.result

    def insert(self, word: str, root: TrieNode) -> None:
        curr = root

        for i in range(len(word)):
            char = word[i]
            if not curr.children[ord(char) - ord("a")]:
                curr.children[ord(char) - ord("a")] = TrieNode()
            
            curr = curr.children[ord(char) - ord("a")]
        
        curr.isEnd = True
    
    def dfs(self, curr, path):
        if len(path) >= len(self.result):
            self.result = "".join(path)
        
        for i in range(25, -1, -1):
            if curr.children[i] and curr.children[i].isEnd:
                le = len(path)
                #action
                path.append(chr(i + ord("a")))
                self.dfs(curr.children[i], path)
                path = path[: le]
        
  
#Time Complexity:; O(N * L), N is number of words, L is average word length.
#Space Compelxity: O(N * L)